#!/usr/bin/python

import rospy
import os
from core.util import read_config, read_csv_DataFrame
from core.ur5_commander import UR5Commander
from gazebo_push_open.cabinet_model import Cabinet
import numpy as np
from core.transforms import rot_z, pose_to_matrix
from gazebo_msgs.msg import ContactsState
from core.gazebo import get_link_pose
import csv
from gazebo_msgs.msg import ContactsState
from rospkg import RosPack
from tf2_ros import Buffer, TransformListener
from utils import *
from tqdm import tqdm

if __name__ == '__main__':
    rospy.init_node('multi_contact_our_handle_node')

    IS_SAVING_RESULTS = False
    IS_SAVING_IMAGES = False
    START_FROM_BEGINNING = False
    method_name = 'our'

    # TF buffer setup
    tf_buffer = Buffer()
    tf_listener = TransformListener(tf_buffer)

    rp = RosPack()
    pkg_path = rp.get_path('path_planning')
    
    # Config
    cfg_path = os.path.join(pkg_path, 'config/config_multi-c_%s_handle_axis_left.yaml' % method_name)
    config = read_config(cfg_path)
    
    # Save/load path for results
    csv_path = config['results_path']

    # Moveit trajectories path
    trajectories_path = config['moveit_trajectories_path']


    # Gazebo simulation launch file (UR5 moveit config launch)
    gazebo_launch_file = config['gazebo_launch_path']

    # Load door configurations
    door_configs_path = config['cabinet_configs_path']
    doors = np.load(door_configs_path)
    num_doors = doors.shape[0]

    # Load screenshots folder
    save_screenshot_path = config['screenshots_path']

    # RVL config path
    rvl_cfg = config['rvl_config_path'] 

    # If False, the data loads and the experiment starts where it stopped
    start_i = 0
    if START_FROM_BEGINNING:
        if IS_SAVING_RESULTS:
            with open(csv_path, 'w') as f:
                writer = csv.writer(f, delimiter=',')
                writer.writerow(['handle_path_exec', 'path_found', 'traj_success', 'contact_free', 'door_opened', 'door_width', 'door_height', 'x', 'y', 'z', 'rot_z', 'state_angle', 'axis_pos'])
    else:
        # data = read_csv_DataFrame(csv_path)
        # start_i = data.shape[0]
        # # doors = doors[rows:, :]
        data = read_csv_DataFrame(csv_path)

        data = data.loc[((data['path_found'] == True) & 
                                (data['traj_success'] == False) & 
                                (data['contact_free'] == True) & 
                                (data['door_opened'] == True))] 
        # data = data.loc[(data['contact_free'] == False)] 
    
    T_G_T = np.load(config['gripper_tool_pose'])
    T_R_W = np.load(config['robot_world_pose'])
    T_W_R = np.linalg.inv(T_R_W)

    # Tool pose w.r.t. handle (it will always be the same)
    axis_pos = -1.0
    Tz45 = np.eye(4)
    Tz45[:3,:3] = rot_z(np.radians(45.))
    T_T_H = np.eye(4)
    T_T_H[:3, :3] = np.array([[0, 0, -axis_pos],
                                [axis_pos, 0, 0],
                                [0, -1, 0]])
    T_T_H = T_T_H @ Tz45
    T_T_H[0, 3] -= 0.28

    # Initial pose in joint space
    q_init = np.array([0., -np.pi*0.5, 0., -np.pi*0.5, 0., 0.])
    q_init_ros = q_init.copy()
    # Adjust joint values from ROS
    q_init[0] += np.pi
    q_init[5] += np.pi
    # q_init[q_init>np.pi]-=(2.0*np.pi)
    # q_init[q_init<-np.pi]+=(2.0*np.pi)

    # Static cabinet params
    door_thickness=config['cabinet_door_dims']['depth']
    static_depth=config['cabinet_door_dims']['static_depth']

    # Static mesh filename
    cabinet_static_mesh_filename = config['cabinet_static_mesh_save_path']
    cabinet_panel_mesh_filename = config['cabinet_panel_mesh_save_path']
    cabinet_full_mesh_filename = config['cabinet_full_mesh_save_path']

    # Generate rotations for all states
    num_states = 30
    opening_angles = np.linspace(0, axis_pos*np.pi*0.5, num=num_states)
    cos_vals = np.cos(opening_angles)
    sin_vals = np.sin(opening_angles)
    Tz_rots = np.zeros((num_states,4,4))
    Tz_rots[:,0,0] = cos_vals
    Tz_rots[:,0,1] = -sin_vals
    Tz_rots[:,1,0] = sin_vals
    Tz_rots[:,1,1] = cos_vals
    Tz_rots[:,2,2] = 1.
    Tz_rots[:,3,3] = 1.

    # Chebyshev joint distance threshold
    qdist_thresh = 0.25 * np.pi

    i = start_i
    # while i < num_doors:
    for i in data.index.values.tolist():

        print('Cabinet %d' % i)
        try:
            handle_path_exec = False
            path_found = False
            trajectory_successful = False
            door_opened = False
            contact_free = True
            final_success = False
            
            # Cabinet parameters
            door = doors[i, :]
            width = door[0]
            height = door[1]
            position = door[2:5]
            rot_z_deg = door[5]
            state_angle = door[6]
            axis_pos = door[7]

            #
            # Load handle opnening traj - if not enough values found, the doors are not open enough
            q_moveit_traj = np.loadtxt(os.path.join(trajectories_path, 'traj_%d.csv' % i), delimiter=',')
            # Ensure it's always a 2D array
            if q_moveit_traj.ndim == 1:  # Only one row, make it 2D
                q_moveit_traj = q_moveit_traj.reshape(1, -1)
            # If nothing other than home position is found or doors are not open enough
            if q_moveit_traj.shape[0] < 2 or abs(np.rad2deg(opening_angles[q_moveit_traj.shape[0] - 2])) < abs(state_angle):
                if IS_SAVING_RESULTS:
                    with open(csv_path, 'a') as f:
                        writer = csv.writer(f, delimiter=',')
                        writer.writerow([handle_path_exec, path_found, trajectory_successful, contact_free, door_opened, width, height, position[0], position[1], position[2], rot_z_deg, state_angle, axis_pos])
                i+=1
                continue


            T_A_S = np.eye(4)
            T_A_S[:3, 3] = np.array(position)
            # T_A_S[2, 3] += T_R_W[2, 3]
            Tz = np.eye(4)
            Tz[:3, :3] = rot_z(np.radians(rot_z_deg))
            T_A_S = T_A_S @ Tz

            # Create a cabinet object
            cabinet_model = Cabinet(door_params=np.array([width, height, door_thickness, static_depth]), 
                                    axis_pos=axis_pos,
                                    T_A_S=T_A_S,
                                    save_path=config['cabinet_urdf_save_path'],
                                    has_handle=True)
                    
            # Save cabinet mesh to a file
            cabinet_model.save_mesh_without_doors(cabinet_static_mesh_filename)
            cabinet_model.save_door_panel_mesh(cabinet_panel_mesh_filename)
            cabinet_model.change_door_angle(state_angle)
            cabinet_model.update_mesh()
            # Save full mesh
            cabinet_model.save_mesh(cabinet_full_mesh_filename)

            # 
            # Start Gazebo processes
            kill_ros_nodes()
            rospy.sleep(1.)
            kill_processes()
            reset_tf_buffer(tf_buffer)
            # gazebo_process = start_gazebo_processes(gazebo_launch_file, tf_buffer, T_R_W)   
            launch_process = subprocess.Popen("source /home/RVLuser/ferit_ur5_ws/devel/setup.bash && roslaunch ur5_robotiq_ft_3f_moveit_config demo_gazebo.launch", shell=True, executable="/bin/bash")
            rospy.sleep(4.)

            # Start the robot commander
            try:
                robot = UR5Commander()
            except RuntimeError as e:
                kill_processes()
                launch_process.terminate()
                launch_process.wait()
                continue

            # Sleep to ensure the robot is ready
            rospy.sleep(1.)

            # 
            # Go to handle grasp pose
            robot.send_multiple_joint_space_poses_to_robot2(q_moveit_traj[:2])
            
            # Spawning model in Gazebo
            cabinet_model.delete_model_gazebo()
            cabinet_model.spawn_model_gazebo()

            # Set contact sub
            contact_state = {'contact_free': True}
            contact_sub = rospy.Subscriber('/contact', ContactsState, contact_callback, contact_state)
            rospy.sleep(1.)

            # Attach gripper to handle
            if not attach_two_models('robot', 'wrist_3_link', 'my_cabinet', 'door_link', 5.0):
                rospy.logerr("Attachment failed.")
                continue # will continue with the same 

            handle_path_exec = robot.send_multiple_joint_space_poses_to_robot2(q_moveit_traj[1:])
            
            if not handle_path_exec:
                contact_free = contact_state['contact_free']

                if IS_SAVING_RESULTS:
                    with open(csv_path, 'a') as f:
                        writer = csv.writer(f, delimiter=',')
                        writer.writerow([handle_path_exec, path_found, trajectory_successful, contact_free, door_opened, width, height, position[0], position[1], position[2], rot_z_deg, state_angle, axis_pos])
                i+=1
                continue
            
            handle_opened_state = np.rad2deg(cabinet_model.get_door_state_gazebo()[1])

            if handle_opened_state >= 85.:
                path_found = True
                trajectory_successful = True
                door_opened = True
                contact_free = contact_state['contact_free']

                if IS_SAVING_RESULTS:
                    with open(csv_path, 'a') as f:
                        writer = csv.writer(f, delimiter=',')
                        writer.writerow([handle_path_exec, path_found, trajectory_successful, contact_free, door_opened, width, height, position[0], position[1], position[2], rot_z_deg, state_angle, axis_pos])
                i+=1
                continue


            # Attach gripper to handle
            if not detach_two_models('robot', 'wrist_3_link', 'my_cabinet', 'door_link', 5.0):
                rospy.logerr("Detachment failed.")
                continue # will continue with the same 
            
            # Delete for the robot to return to home position
            cabinet_model.delete_model_gazebo()

            q_return = [robot.get_current_joint_values(), q_init_ros.tolist()]
            robot.send_multiple_joint_space_poses_to_robot2(q_return)

            # RVL path planning
            T_G_0_array, q, all_feasible_paths, all_feasible_paths_q = rvl_path_planning(rvl_cfg, T_R_W, q_init, 17, False, cabinet_model, handle_opened_state)

            if T_G_0_array.shape[0] == 1:
                print('Path not found')
                if IS_SAVING_RESULTS:
                    with open(csv_path, 'a') as f:
                        writer = csv.writer(f, delimiter=',')
                        writer.writerow([i, path_found, trajectory_successful, contact_free, door_opened, width, height, position[0], position[1], position[2], rot_z_deg, state_angle, axis_pos])
                i+=1
                continue

            path_found = True

            # Points from the path planner
            # adjust joint values to ROS
            q[:, 0] -= np.pi
            q[:, 5] -= np.pi
            q[q>np.pi]-=(2.0*np.pi)     
            q[q<-np.pi]+=(2.0*np.pi)
            q = np.unwrap(q, axis=0)
            q = q.tolist()


            # # Spawn the cabinet in Moveit
            # robot.add_mesh_to_scene(cabinet_full_mesh_filename, 'cabinet', cabinet_model.T_A_S)

            # Go to the first point
            robot.send_multiple_joint_space_poses_to_robot2(q[:2])

            # Spawning model in Gazebo
            cabinet_model.delete_model_gazebo()
            cabinet_model.spawn_model_gazebo()
            rospy.sleep(1.)
            
            # Open doors in Gazebo
            cabinet_model.set_door_state_gazebo(handle_opened_state)
            cabinet_model.change_door_angle(handle_opened_state)
            rospy.sleep(1.)


            # Execute the trajectory and wait until it finishes
            trajectory_successful = robot.send_multiple_joint_space_poses_to_robot2(q[1:])

            cabinet_final_pose = get_link_pose('my_cabinet', 'door_link')
            T_A_S_final = pose_to_matrix(cabinet_final_pose)
            dist = np.linalg.norm(T_A_S_final[:3, 3] - T_A_S[:3, 3])

            final_door_state = np.rad2deg(cabinet_model.get_door_state_gazebo()[1])
            print('Door angle: %f' % final_door_state)
            door_opened = 85.0 <= abs(final_door_state) <= 95.0

            # Check if there was contact
            contact_free = contact_state['contact_free']
            
            if trajectory_successful and contact_free and door_opened:
                print('Experiment finished successfully')
                final_success = True

            # Save end results
            if IS_SAVING_RESULTS:
                with open(csv_path, 'a') as f:
                    writer = csv.writer(f, delimiter=',')
                    writer.writerow([handle_path_exec, path_found, trajectory_successful, contact_free, door_opened, width, height, position[0], position[1], position[2], rot_z_deg, state_angle, axis_pos])
            
            # Shutdown Gazebo simulation and kill all of its processes
            # stop_gazebo_launcher(gazebo_process)
            launch_process.terminate()
            launch_process.wait()
            kill_processes()
            
            #
            i += 1


        except rospy.exceptions.ROSTimeMovedBackwardsException as e:
            rospy.logwarn(f"Time moved backwards. Exception: {e}")
            rospy.loginfo(f"Current ROS time: {rospy.Time.now()}")
            rospy.sleep(2)  # Let time stabilize


